### PR TITLE
Maxlength validations added to the free text fields

### PR DESCRIPTION
**Pull request by NikhilNanjappa**
- Created 3 maxlengths variables which are small(40), medium(80), large(255)
- Currently passing the appropriate maxlengths to the error messages
- Welsh equivalent of the messages added

Used https://github.com/biglotteryfund/salesforce-config/blob/master/data/formmapping/A4AV1Mapping.md for validation specs.

### CODE REVIEW

**Comment by NikhilNanjappa** [comment]
Hi @scottoakley - Could you please go through all the error messages texts and confirm if it looks okay or needs to be changed to something as per your advice.

I got the format from GDS design pattern (https://design-system.service.gov.uk/components/character-count/#error-messages).

eg. `Some field must be 80 characters or less`.

**Comment by scottoakley** [comment]
Hey @davidrapson and @NikhilNanjappa - unfortunately I'm about to finish up for a week's holiday in 15 mins, but I'd like to spend some time properly looking at the error messages before they go live in the form. And I don't think Lily will have time to review them with everything she has on next week. Is it possible to hold off till I'm back?

File: controllers/apply/awards-for-all/fields.js [hunk 1]
```diff
@@ -160,6 +161,22 @@ module.exports = function fieldsFor({ locale, data = {} }) {
                         cy: ''
                     })
                 },
+                {
+                    type: 'string.max',
+                    key: 'line1',
+                    message: localise({
+                        en: `The building and street must be ${FREE_TEXT_MAXLENGTH.large} characters or less`,
```
Comment (scottoakley) [inline controllers/apply/awards-for-all/fields.js]: `Building and street must be ${FREE_TEXT_MAXLENGTH.large} characters or less`

File: controllers/apply/awards-for-all/fields.js [hunk 2]
```diff
@@ -160,6 +161,22 @@ module.exports = function fieldsFor({ locale, data = {} }) {
...
+                        cy: ''
+                    })
+                },
+                {
+                    type: 'string.max',
+                    key: 'line2',
+                    message: localise({
+                        en: `The address line must be ${FREE_TEXT_MAXLENGTH.large} characters or less`,
```
Comment (scottoakley) [inline controllers/apply/awards-for-all/fields.js]: `Address line must be ${FREE_TEXT_MAXLENGTH.large} characters or less`

File: controllers/apply/awards-for-all/fields.js [hunk 3]
```diff
@@ -227,6 +260,22 @@ module.exports = function fieldsFor({ locale, data = {} }) {
                         cy: ''
                     })
                 },
+                {
+                    type: 'string.max',
+                    key: 'line1',
+                    message: localise({
+                        en: `The building and street must be ${FREE_TEXT_MAXLENGTH.large} characters or less`,
```
Comment (scottoakley) [inline controllers/apply/awards-for-all/fields.js]: `Building and street must be ${FREE_TEXT_MAXLENGTH.large} characters or less`

File: controllers/apply/awards-for-all/fields.js [hunk 4]
```diff
@@ -227,6 +260,22 @@ module.exports = function fieldsFor({ locale, data = {} }) {
...
+                        cy: ''
+                    })
+                },
+                {
+                    type: 'string.max',
+                    key: 'line2',
+                    message: localise({
+                        en: `The address line must be ${FREE_TEXT_MAXLENGTH.large} characters or less`,
```
Comment (scottoakley) [inline controllers/apply/awards-for-all/fields.js]: `Address line must be ${FREE_TEXT_MAXLENGTH.large} characters or less`

File: controllers/apply/awards-for-all/fields.js [hunk 5]
```diff
@@ -829,6 +910,13 @@ module.exports = function fieldsFor({ locale, data = {} }) {
                         en: 'Enter your organisation’s Companies House number',
                         cy: ''
                     })
+                },
+                {
+                    type: 'string.max',
+                    message: localise({
+                        en: `Organisation’s Companies House number must be ${FREE_TEXT_MAXLENGTH.large} characters or less`,
```
Comment (scottoakley) [inline controllers/apply/awards-for-all/fields.js]: `Companies House number must be ${FREE_TEXT_MAXLENGTH.large} characters or less`

File: controllers/apply/awards-for-all/fields.js [hunk 6]
```diff
@@ -867,6 +955,13 @@ module.exports = function fieldsFor({ locale, data = {} }) {
                         en: 'Enter your organisation’s charity number',
                         cy: ''
                     })
+                },
+                {
+                    type: 'string.max',
+                    message: localise({
+                        en: `Organisation’s charity number must be ${FREE_TEXT_MAXLENGTH.large} characters or less`,
```
Comment (scottoakley) [inline controllers/apply/awards-for-all/fields.js]: `Charity registration number must be ${FREE_TEXT_MAXLENGTH.large} characters or less`

File: controllers/apply/awards-for-all/fields.js [hunk 7]
```diff
@@ -890,6 +984,13 @@ module.exports = function fieldsFor({ locale, data = {} }) {
                         en: `Enter your organisation’s Department for Education number`,
                         cy: ''
                     })
+                },
+                {
+                    type: 'string.max',
+                    message: localise({
+                        en: `Organisation’s Department for Education number must be ${FREE_TEXT_MAXLENGTH.large} characters or less`,
```
Comment (scottoakley) [inline controllers/apply/awards-for-all/fields.js]: 'Department for Education number must be ${FREE_TEXT_MAXLENGTH.large} characters or less`

File: controllers/apply/awards-for-all/fields.js [hunk 8]
```diff
@@ -1032,12 +1133,19 @@ module.exports = function fieldsFor({ locale, data = {} }) {
...
             messages: [
                 {
                     type: 'base',
                     message: localise({ en: 'Enter a project name', cy: '' })
-                }
+                },
+                {
+                    type: 'string.max',
+                    message: localise({
+                        en: `The project name must be ${FREE_TEXT_MAXLENGTH.medium} characters or less`,
```
Comment (scottoakley) [inline controllers/apply/awards-for-all/fields.js]: `Project name must be ${FREE_TEXT_MAXLENGTH.medium} characters or less`

File: controllers/apply/awards-for-all/fields.js [hunk 9]
```diff
@@ -1099,6 +1204,13 @@ module.exports = function fieldsFor({ locale, data = {} }) {
                         en: `Tell us the towns, villages or wards your beneficiaries live in`,
                         cy: ''
                     })
+                },
+                {
+                    type: 'string.max',
+                    message: localise({
+                        en: `The Project location description must be ${FREE_TEXT_MAXLENGTH.large} characters or less`,
```
Comment (scottoakley) [inline controllers/apply/awards-for-all/fields.js]: `Project locations must be ${FREE_TEXT_MAXLENGTH.large} characters or less`

File: controllers/apply/awards-for-all/fields.js [hunk 10]
```diff
@@ -1394,10 +1514,19 @@ module.exports = function fieldsFor({ locale, data = {} }) {
...
+                    .max(FREE_TEXT_MAXLENGTH.large)
                     .optional(),
                 otherwise: Joi.any().strip()
             }),
-            messages: []
+            messages: [
+                {
+                    type: 'string.max',
+                    message: localise({
+                        en: `The Beneficiaries specific groups must be ${FREE_TEXT_MAXLENGTH.large} characters or less`,
```
Comment (scottoakley) [inline controllers/apply/awards-for-all/fields.js]: Is this for the 'Other' field on the **Specific groups of people (Step 2 of 9)** page? If so: 

'Other specific groups must be ${FREE_TEXT_MAXLENGTH.large} characters or less`

File: controllers/apply/awards-for-all/fields.js [hunk 11]
```diff
@@ -1735,8 +1864,17 @@ module.exports = function fieldsFor({ locale, data = {} }) {
...
             schema: Joi.string()
                 .allow('')
+                .max(FREE_TEXT_MAXLENGTH.large)
                 .optional(),
-            messages: []
+            messages: [
+                {
+                    type: 'string.max',
+                    message: localise({
+                        en: `The Religion or belief must be ${FREE_TEXT_MAXLENGTH.large} characters or less`,
```
Comment (scottoakley) [inline controllers/apply/awards-for-all/fields.js]: Is this for the 'Other' field on the **Religion or belief (Step 7 of 9)** page? If so:

`Other religions or beliefs must be ${FREE_TEXT_MAXLENGTH.large} characters or less`

File: controllers/apply/awards-for-all/fields.js [hunk 12]
```diff
@@ -1780,6 +1919,13 @@ module.exports = function fieldsFor({ locale, data = {} }) {
                         en: `Select the amount of people who speak Welsh that will benefit from your project`,
                         cy: ''
                     })
+                },
+                {
+                    type: 'string.max',
+                    message: localise({
+                        en: `The beneficiaries must be ${FREE_TEXT_MAXLENGTH.large} characters or less`,
```
Comment (scottoakley) [inline controllers/apply/awards-for-all/fields.js]: I'm not sure what field this is for - @NikhilNanjappa can you please clarify?

File: controllers/apply/awards-for-all/fields.js [hunk 13]
```diff
@@ -1852,14 +1998,21 @@ module.exports = function fieldsFor({ locale, data = {} }) {
...
                     type: 'base',
                     message: localise({
                         en: 'Enter the full legal name of the organisation',
                         cy: ''
                     })
+                },
+                {
+                    type: 'string.max',
+                    message: localise({
+                        en: `Full legal name of the organisation must be ${FREE_TEXT_MAXLENGTH.large} characters or less`,
```
Comment (scottoakley) [inline controllers/apply/awards-for-all/fields.js]: 'Full legal name of organisation must be ${FREE_TEXT_MAXLENGTH.large} characters or less`

File: controllers/apply/awards-for-all/fields.js [hunk 14]
```diff
@@ -1873,8 +2026,17 @@ module.exports = function fieldsFor({ locale, data = {} }) {
...
             schema: Joi.string()
                 .allow('')
+                .max(FREE_TEXT_MAXLENGTH.large)
                 .optional(),
-            messages: []
+            messages: [
+                {
+                    type: 'string.max',
+                    message: localise({
+                        en: `The organisation trading name must be ${FREE_TEXT_MAXLENGTH.large} characters or less`,
```
Comment (scottoakley) [inline controllers/apply/awards-for-all/fields.js]: `Organisation's day-to-day name must be ${FREE_TEXT_MAXLENGTH.large} characters or less`

File: controllers/apply/awards-for-all/fields.js (hunk 12 above)
Comment (mattandrews) [reply controllers/apply/awards-for-all/fields.js]: It's a list of people who'll speak Welsh (half, less than half, etc) which probably doesn't need this validation as it's not free text...?

Comment (scottoakley) [reply controllers/apply/awards-for-all/fields.js]: Thanks Matt - that's what I was thinking too.

File: controllers/apply/awards-for-all/fields.js [hunk 15]
```diff
@@ -2506,14 +2702,21 @@ module.exports = function fieldsFor({ locale, data = {} }) {
...
                     type: 'base',
                     message: localise({
                         en: `Enter the full name of the person completing this form`,
                         cy: ''
                     })
+                },
+                {
+                    type: 'string.max',
+                    message: localise({
+                        en: `Full name of the person must be ${FREE_TEXT_MAXLENGTH.large} characters or less`,
```
Comment (scottoakley) [inline controllers/apply/awards-for-all/fields.js]: `Full name must be ${FREE_TEXT_MAXLENGTH.large} characters or less`

File: controllers/apply/awards-for-all/fields.js [hunk 16]
```diff
@@ -2522,14 +2725,21 @@ module.exports = function fieldsFor({ locale, data = {} }) {
...
                     type: 'base',
                     message: localise({
                         en: `Enter the position of the person completing this form`,
                         cy: ''
                     })
+                },
+                {
+                    type: 'string.max',
+                    message: localise({
+                        en: `Position of the person must be ${FREE_TEXT_MAXLENGTH.large} characters or less`,
```
Comment (scottoakley) [inline controllers/apply/awards-for-all/fields.js]: `Position in organisation must be ${FREE_TEXT_MAXLENGTH.large} characters or less`

File: controllers/apply/awards-for-all/fields.js (hunk 10 above)
Comment (davidrapson) [reply controllers/apply/awards-for-all/fields.js]: Yep, spot on

**Comment by NikhilNanjappa** [comment]
Great work @scottoakley ! Thanks a lot for reviewing all the messages quickly (which were more like placeholders to be honest), really appreciate it.

Will work on the changes and give you a shout if any questions.

**Comment by davidrapson** [comment]
[Implement max character limits for all text fields](https://trello.com/c/9zSfQ0Wb/737-implement-max-character-limits-for-all-text-fields)

File: controllers/apply/awards-for-all/fields.js (hunk 12 above)
Comment (NikhilNanjappa) [reply controllers/apply/awards-for-all/fields.js]: My bad .. I will remove this validation from here.

**Comment by NikhilNanjappa** [comment]
@scottoakley - Have made the suggested changes. Could you please have a quick look if everything is as suggested. 

@mattandrews , @davidrapson - I have also sent over an email to Osian requesting these new lines to be translated. I'll implement those changes as soon as I hear from him.

**Comment by davidrapson** [comment]
👍 Fab work, add the welsh translations when you have them and then 🚀 ship it!

**Comment by NikhilNanjappa** [comment]
Got the Welsh translations back from Osian. I have added them in.

**Pull request merged**